lib/worker/OperationQueue.js

```diff
@@ -0,0 +1,363 @@
+      const operation = this.queue[i];
+      if(opCount >= this.MAX_OPERATIONS) {
+        break;
+      }
+      if(head === null) {
+        try {
+          const {valid} = await this._isValidOperation({operation});
+          if(!valid) {
+            this.removedOperations.push(operation);
+            logger.debug(OPERATION_VALIDATION_ERROR, {operation});
+            continue;
+          }
+          head = i;
+        } catch(e) {
+          this.removedOperations.push(operation);
+          logger.debug(OPERATION_VALIDATION_ERROR,
+            {ledgerNodeId: this.ledgerNodeId, operation, error: e});
+          continue;
+        }
+      }
+      ++opCount;
+    }
+
+    if(this.removedOperations.length > 0) {
+      this.removing = this._removeOperationsFromQueue().catch(e => {
+        logger.error(OPERATIONS_REMOVAL_ERROR,
+          {ledgerNodeId: this.ledgerNodeId, error: e});
+      });
+    }
+
+    return {head, opCount};
+  }
+
+  async _isDatabaseEmpty() {
+    const {ledgerNodeId} = this;
+    const projection = {_id: 1};
+    try {
+      const count = await this.collection.find({ledgerNodeId}, {projection})
+        .limit(1).count();
+      return count === 0;
+    } catch(e) {
+      logger.error('Error checking database for empty operation queue.', {
+        ledgerNodeId,
+        error: e
+      });
+      return true;
+    }
+  }
+
+  async _removeOperationsFromQueue({operations = []} = {}) {
+    for(const operation of operations) {
+      this.removedOperations.push(operation);
+    }
+
+    if(this.removedOperations.length === 0) {
+      return;
+    }
+
+    const {ledgerNodeId} = this;
+    try {
+      const bulkOp = this.collection.initializeUnorderedBulkOp();
+
+      for(const {meta} of this.removedOperations) {
+        bulkOp.find({
+          ledgerNodeId,
+          'meta.operationHash': meta.operationHash
+        }).delete();
+      }
+
+      await bulkOp.execute();
+
+      this.removedOperations = [];
+    } catch(e) {
+      logger.error(OPERATIONS_REMOVAL_ERROR, {ledgerNodeId, error: e});
+    }
+  }
+
+  async _isValidOperation({operation}) {
+    const {meta: {operationHash}} = operation;
+    const {ledgerNode} = this;
+
+    const [result, exists] = await Promise.all([
+      this._validateOperation({
+        basisBlockHeight: this.worker.consensusState.blockHeight,
+        ledgerNode,
+        operation
+      }),
+      ledgerNode.storage.operations.exists({operationHash})
+    ]);
+
+    return {valid: !exists && result.valid};
+  }
+
+  async _validateOperation({operation, basisBlockHeight}) {
+    const {meta: {basisBlockHeight: operationBbh}} = operation;
+    const {ledgerNode} = this;
+
+    // Operations are validated at the latest block height of the ledger node
+    // before being accepted into the queue. We can return early when the ledger
+    // node has not advanced its block height, therefore preventing
+    // duplicating the execution of the same validation check.
```

Comment:
When using full sentences/paragraphs, etc. use multiline comment syntax.

```suggestion
    /* Operations are validated at the latest block height of the ledger node
    before being accepted into the queue. We can return early when the ledger
    node has not advanced its block height, therefore preventing
    duplicating the execution of the same validation check. */
```